File: packages/sdk-core/scripts/verify_test_vector.mjs
/**
 * Verify note_commitment_v1.json against sdk-core.
 */
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import {
  createCircomlibPoseidon,
  computeCommitment,
  computeNullifier,
  deriveViewKey,
  computeAttestationDigest,
  ATTESTATION_KIND,
} from "../dist/index.js";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const vectorPath = path.join(__dirname, "../test-vectors/note_commitment_v1.json");

async function main() {
  const doc = JSON.parse(fs.readFileSync(vectorPath, "utf8"));
  if (!doc.expected?.commitment || !doc.expected?.nullifierLeaf0) {
    throw new Error("expected.* missing — run generate_test_vector.mjs first");
  }
  const p = doc.preimage;
  const poseidon = await createCircomlibPoseidon();
  const note = {
    version: BigInt(p.version),
    assetId: BigInt(p.assetId),
    value: BigInt(p.value),
    spendingKey: BigInt(p.spendingKey),
    nullifierKey: BigInt(p.nullifierKey),
    blinding: BigInt(p.blinding),
  };
  const commitment = await computeCommitment(note, poseidon);
  const nullifierLeaf0 = await computeNullifier(
    note.nullifierKey,
    commitment,
    0,
    poseidon
  );
  const viewKey = await deriveViewKey(note.spendingKey, note.nullifierKey, poseidon);
  const attestationDigest = computeAttestationDigest({
    kind: ATTESTATION_KIND.valueBoundDev,
    commitment: commitment.toString(),
    assetId: note.assetId.toString(),
    extra: "100000",
    audienceTag: "1",
  });

  const ok =
    commitment.toString() === doc.expected.commitment &&
    nullifierLeaf0.toString() === doc.expected.nullifierLeaf0 &&
    (!doc.expected.viewKey || viewKey.toString() === doc.expected.viewKey) &&
    (!doc.expected.attestationDigestValueBoundSample ||
      attestationDigest === doc.expected.attestationDigestValueBoundSample);

  console.log(
    JSON.stringify(
      {
        ok,
        client: "sdk-core",
        commitment: commitment.toString(),
        nullifierLeaf0: nullifierLeaf0.toString(),
        viewKey: viewKey.toString(),
        attestationDigest,
        expectedCommitment: doc.expected.commitment,
        expectedNullifierLeaf0: doc.expected.nullifierLeaf0,
        expectedViewKey: doc.expected.viewKey ?? null,
      },
      null,
      2
    )
  );
  if (!ok) process.exitCode = 1;
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: packages/python-client/scripts/verify_test_vector.py
"""Verify note_commitment_v1.json against Python Poseidon (+ view key)."""

from __future__ import annotations

import json
import sys
from pathlib import Path

SCRIPT_DIR = Path(__file__).resolve().parent
PKG_ROOT = SCRIPT_DIR.parent
if str(PKG_ROOT) not in sys.path:
    sys.path.insert(0, str(PKG_ROOT))

from absolute_privacy.note import compute_commitment, compute_nullifier
from absolute_privacy.view_key import derive_view_key

VECTOR = PKG_ROOT.parent / "sdk-core" / "test-vectors" / "note_commitment_v1.json"


def main() -> int:
    doc = json.loads(VECTOR.read_text(encoding="utf-8"))
    expected = doc.get("expected") or {}
    if not expected.get("commitment") or not expected.get("nullifierLeaf0"):
        print("expected.* missing — generate via sdk-core first", file=sys.stderr)
        return 1

    p = doc["preimage"]
    commitment = compute_commitment(
        version=int(p["version"]),
        asset_id=int(p["assetId"]),
        value=int(p["value"]),
        spending_key=int(p["spendingKey"]),
        nullifier_key=int(p["nullifierKey"]),
        blinding=int(p["blinding"]),
    )
    nullifier = compute_nullifier(int(p["nullifierKey"]), commitment, 0)
    view_key = derive_view_key(int(p["spendingKey"]), int(p["nullifierKey"]))

    ok = str(commitment) == str(expected["commitment"]) and str(nullifier) == str(
        expected["nullifierLeaf0"]
    )
    if expected.get("viewKey") is not None:
        ok = ok and str(view_key) == str(expected["viewKey"])

    print(
        json.dumps(
            {
                "ok": ok,
                "client": "python",
                "commitment": str(commitment),
                "nullifierLeaf0": str(nullifier),
                "viewKey": str(view_key),
                "expectedCommitment": expected["commitment"],
                "expectedNullifierLeaf0": expected["nullifierLeaf0"],
                "expectedViewKey": expected.get("viewKey"),
            },
            indent=2,
        )
    )
    return 0 if ok else 1


if __name__ == "__main__":
    raise SystemExit(main())
